README.md
# ab2ascii
Converts AmigaBasic Source-Code-Files to ASCII without orginal AmigaBasic-Installation

ab2ascii reads binary AmigaBASIC sources, expands tokens andsymbols and generates ASCII output.  
Today there is need for ab2ascii since AmigaBASIC does not run anymore under versions of Kickstart V37+.  
ab2ascii should do exactly the same as the AmigaBASIC equivalent

                            SAVE "filename",A

In fact many interresting AmigaBASIC programs have been published in the past and they should not be lost.

By default ab2ascii converts from the standard input stream to the stdout, i.e. you can convert your binaries through a pipe.  
If any input filenames are given in the command line these will be used instead of stdin.



CHANGES
AmigaBASIC->ASCII revision history
(current version: 2.0)

--- May 2020
Port to Linux and change of the Makefile to handle AmigaDOS and Linux
Port to Windows, add Visual Studio 2017 project file

--- Jan 1994
Complete revision of the original ab2ascii by Stefan Reisner.
Several bug fixes in the code symbol handling.
Now using lookup tables to expand the tokens.
Conversion will now be performed in one single pass.  (This
makes it possible to convert through the pipe.)
Recognition of code line numbering implemented.

Now spread under terms of the GNU Public General License!

--- before Jan 1994
Stefan Reisner
sr@ph-cip.uni-koeln.de
Public Domain


abasic.h
/*
 *  ABASIC.H
 */

/*
 * (c)Copyright 1994 by Tobias Ferber.
 *
 * This file is part of AmigaBASIC->ASCII.
 *
 * AmigaBASIC->ASCII is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 1 of the
 * License, or (at your option) any later version.
 *
 * AmigaBASIC->ASCII is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with AmigaBASIC->ASCII; see the file COPYING.  If not, write to
 * the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 */

#ifndef ABASIC_H

#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#ifdef __linux__
#include <string.h>
#endif

#ifdef __GNUC__
/* suggested parentheses around assignment used as truth value */
#define if(assignment) if( (assignment) )
#define while(assignment) while( (assignment) )
#endif

#define NIL(type)     (type)0L

/* AmigaBASIC variable & label symbols */
typedef struct symbol
{
  struct symbol *next;
  int id;      /* coded in 2 bytes */
  char *s;     /* length coded in 1 byte */

} symbol_t;

#define SYMBOL_SIZE  sizeof(symbol_t)

/* AmigaBASIC code lines */
typedef struct codeline
{
  struct codeline *next;
  int len, flags;
  unsigned char *s;

} codeline_t;

#define CODELINE_SIZE  sizeof(codeline_t)


/*** / GLOBALS / ***/

/* tokens.c */
extern int tok_min, tok_max;  /* 1 byte tokens */
extern int f8_min, f8_max,    /* 2 byte tokens */
           f9_min, f9_max,
           fa_min, fa_max,
           fb_min, fb_max;

extern char *tok_tab[], *f8_tab[], *f9_tab[], *fa_tab[], *fb_tab[];

/* main.c */
extern symbol_t *sym_tab;         /* gloabal symbol table */
extern int numsyms;               /* #of available symbols */
extern codeline_t *code_buffer;   /* list of code lines */
extern long numlines;             /* #of code lines */
extern long current_line;
extern char *whoami;              /* global copy of argv[0] */
extern char *infile;              /* current input filename */
extern FILE *fin, *fout, *ferr;

/*** / PROTOTYPES / ***/

#ifndef __P

#if defined (__STDC__) || defined(__cplusplus)
#define __P(protos) protos
#else /* !(__STDC__ || __cplusplus) */
#define __P(protos) ()
#endif /* __STDC__ || __cplusplus */

#endif /* !__P */

/* main.c */
extern void warn __P( (const char *fmt, ...) );

/* args.c */
extern char *convert_args __P( (char *) );
extern void display_args __P( (void) );

/* flist.c */
extern int chain_fname __P( (char *fname) );
extern char *unchain_fname __P( (void) );
extern void purge_flist __P( (void) );

/* symbols.c */
extern int read_sym __P( (FILE *fp, int len, int id) );
extern char *get_sym __P( (int id) );
extern void free_symbols __P( (void) );

/* codelines.c */
extern int read_line __P( (FILE *fp, int len, int flags) );
extern void free_code __P( (void) );

/* expand.c */
extern void expand_code __P( (void) );

#endif /* !ABASIC_H */


main.c
/*
    AmigaBASIC -> ASCII  -- generate ASCII source from AmigaBASIC binaries

    Original version by Stefan Reisner, sr@ph-cip.uni-koeln.de
    Revised & bug-fixes by Tobias Ferber, ukjg@dkauni2.bitnet
*/

/*
 * (c)Copyright 1994 by Tobias Ferber.
 *
 * This file is part of AmigaBASIC->ASCII.
 *
 * AmigaBASIC->ASCII is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 1 of the
 * License, or (at your option) any later version.
 *
 * AmigaBASIC->ASCII is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with AmigaBASIC->ASCII; see the file COPYING.  If not, write to
 * the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 */

#include "abasic.h"

/* include the version tag */
#include "version.h"

/* include string operations*/
#include "string.h"

/* globals */

/* the symbol table */
symbol_t *sym_tab= NIL(symbol_t *);
int numsyms;

/* the list of code lines */
codeline_t *code_buffer= NIL(codeline_t *);
long current_line, numlines;

char *whoami; /* argv[0] */
char *infile;
int debuglevel= 0;

/* input, output and error stream */
FILE *fin, *fout, *ferr;

static void display_version_information(void)
{
  static char license[]=
    "AmigaBASIC->ASCII is free software; you can redistribute it and/or\n"
    "modify it under the terms of the GNU General Public License as\n"
    "published by the Free Software Foundation; either version 1 of the\n"
    "License, or (at your option) any later version.\n"
    "\n"
    "AmigaBASIC->ASCII is distributed in the hope that it will be useful,\n"
    "but WITHOUT ANY WARRANTY; without even the implied warranty of\n"
    "MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the\n"
    "GNU General Public License for more details.\n"
    "\n"
    "You should have received a copy of the GNU General Public License\n"
    "along with AmigaBASIC->ASCII; see the file COPYING.  If not, write to the\n"
    "Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.\n"
    ;

  puts("AmigaBASIC->ASCII Version " VERSION " (compiled " __DATE__ ", " __TIME__ ")\n"
       "(c)Copyright 1994 by Tobias Ferber, ukjg@dkauni2.bitnet\n"
	  "Linux port 2020 by Thomas Ronzon, thomas@ronzon.de\n");

  puts(license);
}


void warn(const char *fmt, ...)
{
  va_list argp;
  va_start(argp,fmt);
  fprintf(ferr,"%s: ",whoami);

  if(infile)
    fprintf(ferr,"%s: ",infile);

  vfprintf(ferr,(char *)fmt,argp);
  fprintf(ferr,"\n");
  fflush(ferr);
  va_end(argp);
}


int dothehardpart( void )
{
  int len, c, n=0;
  int faux_pas= 0;
  long codesize= 0;

  /* initialize globals */

  numsyms= numlines= 0;

  /* begin scanning the input stream */

  c= fgetc(fin);

  if(c != 0xf5)
  {
    warn("Hmmm... this doesn't look like an AmigaBASIC binary.\n"
         "My output will probably look a bit funny.  Proceed, for a laugh.");
  }

  if( feof(fin) || ferror(fin) )
  { warn("unexpected end of input -- no code, no symbols");
    return 1;
  }

  /* read the code segment */

  do {

    /* Stefan Reisner thought that AmigaBASIC encodes the length of a line
     * in a 2 byte word prefix.  This is not the whole truth...
     * I've experienced that the first byte holds a flag (e.g. 0x80 indicates
     * a line number) and only the second byte encodes the length.
     */

    c= fgetc(fin); /* flags */

    if( len= fgetc(fin) )
    {
      codesize += len;
      len-= 2;

      if( !feof(fin) && !ferror(fin) )
        n= read_line(fin,len,c);
    }

  } while( len && n == len && !ferror(fin) && !feof(fin) );

  if( len )
  {
    if( n != len || ferror(fin) )
    {
      warn("I'm confused; after having read the code segment, I still seem\n"
           "to want to go on.  In fact I just wanted to read another %d bytes\n"
           "when a serious problem interrupted me.  I'll try to forget about this.",len);
    }

    if( feof(fin) )
    {
      warn("Uh, oh.  I suspect there was a null missing at the end of the code segment,\n"
           "causing me to read past where I should stop.\n"
           "I'll forget about the symbols now.  Proceed, with fingers crossed.");
    }
  }
  else if( feof(fin) || ferror(fin) ) /* and len <= 0 */
    warn("This seems to be a funny program.  The input ended without any symbol definitions.\n");

  /* read the symbols */

  if( !ferror(fin) && !feof(fin) )
  {
    int ok= 1;

#ifdef DEBUG
    if(debuglevel >= 1)
      fprintf(stderr,"skipping a '\\x%02x' character\n",fgetc(fin));
    else
#endif
      (void)fgetc(fin);  /* skip null byte */

    if( (codesize & 1) == 0 )
    {
#ifdef DEBUG
      if(debuglevel >= 1)
        fprintf(stderr,"skipping another '\\x%02x' character\n",fgetc(fin));
      else
#endif
      (void)fgetc(fin);
    }

    if( feof(fin) || ferror(fin) )
    {
      warn("I was just about to read the symbols when I reached the end of file.\n"
           "Maybe we have a BASIC program without symbols?  We'll see...");
    }

    else do {
      len= fgetc(fin);

      if( !feof(fin) )
        ok= read_sym(fin, len, numsyms++);

    } while( ok && !feof(fin) && !ferror(fin) );

    if( !ok )
    {
      warn("After having read %ld %s I couldn't get another %d bytes.\n"
           "Maybe there is not enough free store?  I'll better stop reading symbols!",
           numsyms-1, (numsyms-1 == 1) ? "symbol": "symbols");
    }
  }

#ifdef DEBUG
    if(debuglevel >= 1)
      fprintf(stderr,"expanding %ld lines of code w/ %d symbols\n",numlines,numsyms);
#endif

  expand_code();

  /* we _must_ call these to become re-entrant */

  free_symbols();
  free_code();

  return faux_pas;
}


/*
   Generate a filename from a format string fmt and the replacement string s for
   the first `%s' in fmt and open this file for write.
   Normally the FILE pointer will be returned, 0 indicates a serious problem.
   This function prints warning messages via warn().
*/

FILE *fmtopen( char *fmt, char *s )
{
  FILE *fp= (FILE *)0L;

  char *fname= (char *)malloc( strlen(s) + strlen(fmt) );

  if(fname)
  {
    sprintf(fname,fmt,s);
    fp= fopen(fname,"w");

    if(!fp)
      warn("can't write to `%s'",fname);
  }
  else warn("out of memory... aaaiiiiiieeeeeeeee!");

  if(fname)
    free(fname);

  return fp;
}


int main( int argc, char **argv )
{
  int badopt= 0;
  int numfiles= 0;
  char *outfmt;

#ifdef _DCC /* Dice */
  expand_args(argc,argv, &argc,&argv);
#endif /* _DCC */

  infile= outfmt= (char *)0L;
  whoami= *argv;

  fin= stdin;
  fout= stdout;
  ferr= stderr;

  while(--argc>0 && !badopt)
  {
    char *arg= *++argv;

    if(*arg=='-')
    {
      if(arg[1]=='-')
        arg= convert_args(*argv);

      switch(*++arg)
      {

/*-d*/  case 'd':
#ifdef DEBUG
          debuglevel= 1;
#else
          //echo("not compiled w/ a symbol DEBUG defined.  No debugging information available -- Sorry.");
	      printf("not compiled w/ a symbol DEBUG defined.  No debugging information available -- Sorry.");
#endif
          break;

/*-E*/  case 'E':
          if(arg[1]) ++arg;
          else arg= (--argc > 0) ? *(++argv) : (char *)0L;

          if(arg && *arg)
          {
            if(ferr != stderr)
            {
              warn("warning: option `%s' has already been seen!",*argv);
              fclose(ferr);
            }

            ferr= fopen(arg,"w");

            if(!ferr)
            {
              warn("can't direct error output to `%s' -- will use stderr",arg);
              ferr= stderr;
            }
          }
          else
          { warn("missing filename after `%s' option",*argv);
            ++badopt;
          }
          break;

/*-?*/  case '?':
/*-h*/  case 'h':
          fprintf(stderr,
            "usage: %s [options] [-o|>] [outfile] [<] [infiles..]\n\n",whoami);

          display_args();
          badopt= 1; /* hack: means exit. */
          break;

/*-o*/  case 'o':
          if(arg[1]) ++arg;
          else arg= (--argc > 0) ? *(++argv) : (char *)0L;

          if(arg && *arg)
          {
            if(outfmt)
              warn("option `%s' has already been seen!",*argv);
            outfmt= arg;
          }
          else
          {
            warn("missing output filename after `%s' option",*argv);
            ++badopt;
          }
          break;

/*-s*/  case 's':
          warn("silent option `%s' is not yet implemented -- sorry.",*argv);
          break;

/*-v*/  case 'v':
          display_version_information();
          badopt= 1; /* hack: means exit. */
          break;

/*??*/  default:
          warn("unrecognized option `%s'",*argv);
          ++badopt;
          break;
      }
    }
    else
    {
      if(arg && *arg)
      {
        if( chain_fname(arg) )
        {
          warn("out of memory... aaaiiiiiieeeeeeeee!");
          ++badopt;
        }
        else ++numfiles;
      }
      else
      {
        warn("command line error: can't parse `%s'",arg);
        ++badopt;
      }
    }
  }


  if(numfiles)
  {
    while( !badopt && (infile= unchain_fname()) )
    {
      if( fin= fopen(infile,"rb") )
      {
        if( fout= outfmt ? fmtopen(outfmt, infile) : stdout )
        {
          badopt= dothehardpart();

          if(fout != stdout)
            fclose(fout);
        }
        else ++badopt; /* fmtopen() has already warned */
        fclose(fin);
      }
      else
      {
        warn("can't access your input file `%s'",infile);
        ++badopt;
      }
    }
    purge_flist();
  }

  else if( !badopt )
  {
    if( fout= outfmt ? fmtopen(outfmt, "stdin") : stdout )
    {
      badopt= dothehardpart();

      if(fout != stdout)
         fclose(fout);
    }
    else ++badopt; /* fmtopen() has already warned */
  }

  if(fin  && fin  != stdin)    fclose(fin);
  if(fout && fout != stdout)   fclose(fout);
  if(ferr && ferr != stderr)   fclose(ferr);

  exit( badopt ? 1:0 );
}
